nixos/modules/services/networking/unbound.nix

```
@@ -84,31 +71,110 @@ in
         description = "Use and update root trust anchor for DNSSEC validation.";
       };
 
-      extraConfig = mkOption {
-        default = "";
-        type = types.lines;
+      settings = mkOption {
+        default = {};
+        type = with types; submodule {
+
+          freeformType = let
+            validSettingsPrimitiveTypes = oneOf [ int str bool float ];
+            validSettingsTypes = oneOf [ validSettingsPrimitiveTypes (listOf validSettingsPrimitiveTypes) ];
+            settingsType = (attrsOf validSettingsTypes);
+          in attrsOf (oneOf [ settingsType (listOf settingsType) ])
+              // { description = ''
+                unbound.conf configuration type. The format consist of an attribute
+                set of settings. Each settings can be either one value, a list of
+                values or an attribute set. The allowed values are integers,
+                strings, booleans or floats.
+              '';
+            };
+
+          options = {
+            remote-control.control-enable = mkOption {
+              type = bool;
+              default = false;
+              description = "";
+            };
+          };
+       };
+        example = literalExample ''
+          {
+            server = {
+              interface = [ "127.0.0.1" ];
+            };
+            forward-zone = [
+              {
+                name = ".";
+                forward-addr = "1.1.1.1@853#cloudflare-dns.com";
+              }
+              {
+                name = "example.org.";
+                forward-addr = [
+                  "1.1.1.1@853#cloudflare-dns.com"
+                  "1.0.0.1@853#cloudflare-dns.com"
+                ];
+              }
+            ];
+            remote-control.control-enable = true;
+          };
+        '';
         description = ''
-          Extra unbound config. See
-          <citerefentry><refentrytitle>unbound.conf</refentrytitle><manvolnum>8
-          </manvolnum></citerefentry>.
+          Declarative Unbound configuration
+          See the <citerefentry><refentrytitle>unbound.conf</refentrytitle>
+          <manvolnum>5</manvolnum></citerefentry> manpage for a list of
+          available options.
         '';
       };
-
     };
   };
 
   ###### implementation
 
   config = mkIf cfg.enable {
 
+    services.unbound.settings = {
+      server = {
+        directory = mkDefault cfg.stateDir;
+        username = cfg.user;
+        chroot = cfg.stateDir;
+        pidfile = ''""'';
+        interface = mkDefault ([ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1"));
+        access-control = mkDefault ([ "127.0.0.0/8 allow" ] ++ (optional config.networking.enableIPv6 "::1/128 allow"));
+        auto-trust-anchor-file = mkIf cfg.enableRootTrustAnchor rootTrustAnchorFile;
```

```suggestion
        auto-trust-anchor-file = mkIf cfg.enableRootTrustAnchor rootTrustAnchorFile;
        tls-cert-bundle = "${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt";
```

Would it make sense to set `tls-cert-bundle` by default? When creating a forwarding zone with `forward-tls-upstream` enabled, this option must be set or else Unbound fails to resolve domains in that zone with the following error:

```
error: ssl handshake failed crypto error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
notice: ssl handshake failed 1.1.1.1 port 853
```

---

It might indeed make sense. I don't really know how other modules are doing regarding this particular issue. I know postfix does this by default. I'd like more opinions about this before adding it.

---

I don't know of other modules but I'd advise to use `security.pki.certificateFiles`. This option contains `cacert` by default but also allows users to blacklist untrusted CA or add custom certificates.

---

> I don't know of other modules but I'd advise to use `security.pki.certificateFiles`. This option contains `cacert` by default but also allows users to blacklist untrusted CA or add custom certificates.

Or should one use `/etc/ssl/certs/ca-certificates.crt` which as per `configuration.nix(5)` is supposed to contain the output CA bundle:

       security.pki.certificateFiles
           A list of files containing trusted root certificates in PEM format. These are concatenated to form /etc/ssl/certs/ca-certificates.crt, which is used by many
           programs that use OpenSSL, such as curl and git.

---

Uhm, the location is read-only and owned by root, so it should still be deterministic.
If unbound doesn't support loading multiple bundles, then, yes this should do.

---

It looks like unbound indeed does not support loading multiple bundles so I set the default value for this to `/etc/ssl/certs/ca-certificates.crt`.